### PR TITLE
Use Sentence Transformers lib instead of Transformers

### PR DESCRIPTION
**Pull request by raam86**
This PR suggests a move to a higher level library (sentence transformers) the behaviour stays the same while increasing performance by about an order of a magnitude. 

This is probably due to the promise of this library to be highly optimized for such workloads, they also claim to use and distribute to all gpus automatically but I haven't tested that yet. 

I also added support for batching in the form of an input array (instead of a single string) this should help with decoupling the transformation work for large workloads if needed.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,37 +1,22 @@
 import os
 from logging import getLogger
 from fastapi import FastAPI, Response, status
-from vectorizer import Vectorizer, VectorInput
-from meta import Meta
+from transformer import Transformer, TextInput, TextsInput
```
Comment (raam86) [inline app.py:4]: The aim of this PR is to simplify the modul by using a higher level library.  A lot of the functionality implemented by this module is already implemented by sentence transformers  (including robust configs)

File: app.py [hunk 2]
```diff
@@ -1,37 +1,22 @@
...
-    global meta_config
-
-    cuda_env = os.getenv("ENABLE_CUDA")
-    cuda_support=False
-    cuda_core=""
+    global encoder
+    model_path = "/./models/trans" #we assume the model was already downloaded
+    device = os.environ.get("DEVICE") # device can be any pytorch device. 
+    # Current implementation should detect cuda devices automatically
+    encoder = Transformer().load_model(model_path, device)
```
Comment (raam86) [inline app.py:16]: `load_model` returns a function we can use in the API

File: app.py [hunk 3]
```diff
@@ -1,37 +1,22 @@
...
 
-    if cuda_env is not None and cuda_env == "true" or cuda_env == "1":
-        cuda_support=True
-        cuda_core = os.getenv("CUDA_CORE")
```
Comment (raam86) [inline app.py:25]: The library promises to handle the different cuda configs automatically, needs to be tested in cuda on the cloud.

File: app.py [hunk 4]
```diff
@@ -1,37 +1,22 @@
...
-        if cuda_core is None or cuda_core == "":
-            cuda_core = "cuda:0"
```
Comment (raam86) [inline app.py:27]: Should I add this logging?

File: bench_vectorizer.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+from vectorizer import Vectorizer, VectorInput
```
Comment (raam86) [inline bench_vectorizer.py:1]: This file is useful for local testing, in order to run the vectorizing process independently from uvicorn.

File: test_app.py [hunk 6]
```diff
@@ -28,24 +29,34 @@ def wait_for_uvicorn_start():
...
+    # optimized for setnencetransformer the model tries its best to autotune
+    # Like
+    # WARNING:root:No sentence-transformers model found with name /Users/raam/.cache/torch/sentence_transformers/distilroberta-base. Creating a new one with MEAN pooling.
+    # Some weights of the model checkpoint at /Users/raam/.cache/torch/sentence_transformers/distilroberta-base were not used when initializing RobertaModel: ['lm_head.bias', 'lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.weight']
+    #- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
+    #- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
+    assert request.param.split("-")[-1] in config["_name_or_path"].split("-")[-1]
```
Comment (raam86) [inline test_app.py:49]: added a simple assertion making sure correct model was downloaded

File: requirements.txt [hunk 7]
```diff
@@ -1,8 +1,4 @@
-tokenizers==0.10.3
-transformers==4.16.2
```
Comment (raam86) [inline requirements.txt:2]: fastapi 0.78.0 fails if headers are not the right content type (e.g application/json) had to add that explicitly to curl

**Comment by kmulka-bloomberg** [comment]
@raam86 The performance promises of this sound great. Anything stopping this from moving forward? It looks like this needs CLA acceptance from you. But, maybe something else is blocking this?

**Comment by bobvanluijt** [comment]
Looping in @antas-marcin who might have more context (mostly because this is an older PR)

**Comment by erikvullings** [comment]
Would be great if the t2v performance could be improved, as I am getting a lot of timeouts when importing and searching at the same time. So is it possible to merge this PR?

**Comment by kl-thamm** [comment]
Did not look into it this ins great detail but came across a related issue. 
Afaik sentence-transformers vecetorizes each text in a list as a whole. This PR calls `encode` with just one string, which sentence-transformers will then just put into a list. 
From my understanding it will then encode this as a whole. If the input is too long, it will not be split into batches or anything but truncate the input (see e.g. https://github.com/UKPLab/sentence-transformers/issues/181#issuecomment-614693712).

Could it be that the gain of speed would come (at least partly) from the fact that the way this is currently implemented would just encode parts of the text?

**Comment by antas-marcin** [comment]
The SentenceTransformer vectorizer is already a part of transformers project. Closing this PR.

**Pull request closed without merging**